Add receiving groups to shipment

coderbyheart opened the pull request:
> Receiving groups should be able to be allocated when creating a shipment, i.e. If a shipment is going from UK to Calais; L’Auberge, CA, CFC, etc should be able to be added from the start so that sending groups can assign boxes/pallets to certain groups.

*Shipments* should have an association to 1 or more groups which this shipments serves.

This will allow to later pre-populate groups in the shipping offer creation when for example selecting a receiving group for a pallet.

See https://www.notion.so/distributeaid/Receiving-groups-should-be-able-to-be-allocated-when-creating-a-shipment-i-e-If-a-shipment-is-goin-a39e91ca0c67427d95f3496c243694e5

- [x] remove the concept of a "route" (this is not needed because shipments have an origin region and a destination region)
- [x] change association to region to n:1 (one shipment is for exactly one destination region)
- [x] add `receivingGroups` to shipment (a list of 1 or more groups)
- [x] add UI for selecting groups when creating / editing shipment
- [x] change offer form to show only groups assigned to the shipment in dropdown for assiging the pallet to a group
- [x] can we have multiple sending and receiving hubs in a shipment? #767 

## Definition of *region*

Example of regions usage are:

- Groups serve a region.
- Groups serve multiple regions.
- Groups can offer to a region.

> A region is defined by:
> 
> 1. a country (e.g. France)
> 2. and a locality that identifies the region (e.g. Calais).

Example regions are:

- France (Dunkirk)
- France (Calais)
- Greece (Islands)

Groups can select which regions they serve.

Admins can also edit groups and make this selection for them.

## association of a *shipment* to a *region*

> A shipment  
> is associated with a route (where a route is from a country of origin [e.g. “UK”] to a destination country [e.g. “France”])  
> and is associated with one (1) region in the destination country (a region is a country plus a locality, e.g. “France” + “Calais”).

That will then allow offering groups to only select receiving groups which serve this region.

coderbyheart committed:
refactor: remove unused open location code

coderbyheart committed:
refactor(group): make country and locality regular property

This converts the primaryLocation field of a group into two
regular, required fields: country and locality.

Locality is used as a more generic term instead of city.

Currently there is no secondary or other location, so
primary is not neccessary to depict the groups location.

Additionally, all groups can be expected to provide a
country and a locality to identify them in combination
with their group name.

Making country a regular field also simplifies the validation,
and we now also use the list of countries.

ghost commented:
### CodeSee Review Map:
[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/distributeaid/shipment-tracker/699/ff4d0034/2fd401a7528d680e6a0ca1c4cabf038a18008316.svg)](https://app.codesee.io/r/reviews?pr=699&src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fshipment-tracker)

Review [in an interactive map](https://app.codesee.io/r/reviews?pr=699&src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fshipment-tracker)

View [more CodeSee Maps](https://app.codesee.io/r/maps?src=https%3A%2F%2Fgithub.com%2Fdistributeaid%2Fshipment-tracker)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

coderbyheart committed:
refactor: merge migrations

coderbyheart committed:
test(frontend): group location format has changed

coderbyheart committed:
fix(create-group): small spacing fix

coderbyheart committed:
WIP: add regions

coderbyheart committed:
feat: make sure ids are unique

for countries, regions and routes

coderbyheart committed:
fix: unset servingRegions if none selected

coderbyheart committed:
fix: set Calais route to France

coderbyheart committed:
fix: clean up

coderbyheart committed:
docs: add TODO

coderbyheart committed:
feat: routes can serve multiple regions

coderbyheart committed:
feat: filter groups by shipment route

TODO: figure out why array is not converted

coderbyheart committed:
fix: make locality of region mandatory

coderbyheart committed:
fix: return serving regions

coderbyheart committed:
test: fix group tests

coderbyheart committed:
fix(group): add defaults

coderbyheart committed:
feat: remove shipment routes

this is not needed because shipments have an origin region and a destination region

coderbyheart committed:
fix: improve rendering of shipment id

coderbyheart committed:
feat: add receivingGroups to Shipment

coderbyheart committed:
test: add missing field

coderbyheart committed:
fix(shipment): offers are not added by admins

coderbyheart committed:
feat: filter groups in pallet selection based on receivingGroups
